proposals/0000-borrow-operator.md

```diff
@@ -0,0 +1,599 @@
+Non-mutating uses of values are always either:
+
+- *borrows*, which just read the low-level information in the value, or
+- *copies*, which create a new, independent copy of that low-level
+  information according to the rules of the type.
+
+Swift decides whether to borrow or copy a value based on how it is used
+and where it comes from. As a general rule, borrowing a value is less
+expensive and is always preferable when possible. The costs of copying
+values does not always matter, but sometimes they can come to dominate a
+time or memory profile, and it becomes important to find ways to avoid
+them. 
+
+In some situations, Swift cannot borrow a value because it will be used
+in an operation that requires an independent value, such as assigning
+it into a variable. These copies generally cannot be avoided by
+transferring ownership of an existing value, such as with the [`consume`
+operator][SE-0366]. This proposal does not help with this.
+
+In other cases, a value could theoretically be borrowed from where it
+is stored in memory, but Swift instead chooses to emit a *conservative
+copy*. In general, Swift does this in order to avoid conflicts if some
+other code modifies that location while the value is in use. If the
+programmer is sure that such a conflict is impossible, borrowing the
+value instead would avoid the copy. However, it can be tricky to
+convince Swift to actually do this today because there is no feature
+to explicitly request it.
+
+Swift does have some features for avoiding implicit copies of borrowed
+values. For example, under [SE-0377][], you can declare a parameter as
+`borrowing`:
+
+```swift
+func total(_ array: borrowing [Int]) -> Int {
+  ...
+}
+```
+
+Within this function, Swift will not implicitly copy the value of this
+parameter. However, this does not prevent the *caller* of this function
+from potentially having copied the value before passing it in:
+
+```swift
+let sum = total(c)
+```
+
+Swift will make a local decision when compiling this call about whether
+it is necessary to conservatively copy the value. In the simple case of
+a local variable, Swift should be able to guarantee that `c` will be
+borrowed by proving that `c` is not modified during the call. But if
+`c` were a global variable, or a class property, this would be difficult,
+and Swift is likely to emit a conservative copy.[^2]
+
+[^2]: The presence of `borrowing` on the function parameter does not
+affect this decision. If it did, adopting `borrowing` would change the
+semantics of all calls to the function, potentially causing them to
+experience conflicts that they did not before. The authors of SE-0377
+justly wanted the modifier to have only local effects.
+
+## Proposed solution
+
+This proposal adds a new `borrow` operator that explicitly requires a
+value to be borrowed for a particular use.
+
+```swift
+// Force the value of c to be borrowed. If c is somehow modified during
+// this call, there may be an error at runtime.
```

Comment:
"error at runtime" here implies that borrows carry runtime costs that do not apply to similar copies.  Since most people's mental model seems to be that "borrow" is "pointer to" with no other overhead, that cost likely needs to be called out a little more explicitly.

Reply:
Ah.  I see you've explored that in the borrows-by-default section at the end.